http-netty/src/main/java/io/micronaut/http/netty/body/AvailableNettyByteBody.java

```diff
@@ -62,6 +64,31 @@ public static ByteBuf toByteBuf(AvailableByteBody body) {
         }
     }
 
+    /**
+     * This is a wrapper around {@link AvailableNettyByteBody#AvailableNettyByteBody(ByteBuf)}
+     * with an extra body length check.
+     *
+     * @param loop The event loop for constructing {@link StreamingNettyByteBody}
+     * @param bodySizeLimits The body size limits to check
+     * @param buf The input buffer
+     * @return The body with the given input buffer, or a {@link StreamingNettyByteBody} with the
+     * appropriate content length error
+     */
+    @NonNull
+    public static CloseableByteBody createChecked(@NonNull EventLoop loop, @NonNull BodySizeLimits bodySizeLimits, @NonNull ByteBuf buf) {
+        // AvailableNettyByteBody does not support exceptions, so if we hit one of the configured
+        // limits, we return a StreamingNettyByteBody instead.
+        if (buf.readableBytes() > bodySizeLimits.maxBodySize() || buf.readableBytes() > bodySizeLimits.maxBufferSize()) {
+            BufferConsumer.Upstream upstream = bytesConsumed -> {
+            };
+            StreamingNettyByteBody.SharedBuffer mockBuffer = new StreamingNettyByteBody.SharedBuffer(loop, bodySizeLimits, upstream);
+            mockBuffer.add(buf); // this will trigger the exception for exceeded body or buffer size
+            return new StreamingNettyByteBody(mockBuffer);
+        } else {
+            return new AvailableNettyByteBody(buf);
```

Comment:
Personally, I would revert the if and return `AvailableNettyByteBody` and avoid redundant else